File: epi_judge_python/tree_level_order.py
from typing import List

from binary_tree_node import BinaryTreeNode
from test_framework import generic_test
from collections import deque


def binary_tree_depth_order(tree: BinaryTreeNode) -> List[List[int]]:
    que = deque()
    dummy = BinaryTreeNode(None, None, None)
    res = []
    if not tree:
        return []
    que.append(tree)
    que.append(dummy)
    lev = []
    while que:
        n = que.popleft()
        if n is dummy:
            res.append(lev)
            lev = []
            if que:
                que.append(dummy)
            continue
        lev.append(n.data)
        if n.left:
            que.append(n.left)
        if n.right:
            que.append(n.right)
    return res


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('tree_level_order.py',
                                       'tree_level_order.tsv',
                                       binary_tree_depth_order))

File: epi_judge_cpp/tree_level_order.cc
#include <memory>
#include <queue>
#include <utility>
#include <vector>

#include "binary_tree_node.h"
#include "test_framework/generic_test.h"
using std::unique_ptr;
using std::vector;

vector<vector<int>>
BinaryTreeDepthOrder(const unique_ptr<BinaryTreeNode<int>> &tree) {
  // using QueElem = std::pair<int, int>; // <data, level>
  std::queue<const BinaryTreeNode<int>*> que;
  // int cur_level = 0;
  vector<vector<int>> result;
  vector<int> nodes;
  const BinaryTreeNode<int> Sentinel(0);
  for (que.push(tree.get()), que.push(&Sentinel); !que.empty();) {
    if (que.front() == &Sentinel) {
      result.push_back(nodes);
      nodes.clear();
      que.pop();
      if (!que.empty()) {
        que.push(&Sentinel);
      }
    } else {
      nodes.push_back(que.front()->data);
      if (que.front()->left) {
        que.push(que.front()->left.get());
      }
      if (que.front()->right) {
        que.push(que.front()->right.get());
      }
      que.pop();
    }
  }

  // auto [data, level] = que.front();
  //   que.pop();
  //   if (level != cur_level) {
  //     result.emplace_back(nodes);
  //     nodes.clear();
  //     cur_level = level;
  //   }
  //   nodes.push_back(data);
  // }
  if (!nodes.empty()) {
    result.push_back(nodes);
  }
  return result;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree"};
  return GenericTestMain(args, "tree_level_order.cc", "tree_level_order.tsv",
                         &BinaryTreeDepthOrder, DefaultComparator{},
                         param_names);
}
